### ISSUE TITLE
Usage of strdup is not C89-compliant

### ISSUE THREAD

**Issue by aaronpuchert** [OSGeo/PROJ#609]
According to the contributing guidelines, section *Programming language*:

> PROJ.4 is developed strictly in ANSI C 89.

But this isn't enforced. (Using `-ansi` or `-std=c89` should do that.)

Throughout the code there are repeated uses of [strdup](http://en.cppreference.com/w/c/experimental/dynamic/strdup), which is only part of [POSIX C since 2008](https://linux.die.net/man/3/strdup). It's not yet part of the official C standard, and certainly not of ANSI C.

**Comment by rouault** [comment]
While this is true, all platforms that matter support strdup() out of the box, so I don't think we should invest much time about that. Probably that changing the guidelines to a somewhat vague "PROJ.4 is developed in C" would be good enough.

**Comment by aaronpuchert** [comment]
Well, MSVC [only](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strdup-wcsdup) supports [_strdup](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strdup-wcsdup-mbsdup), and warns

> warning C4996: 'strdup': The POSIX name for this item is deprecated. Instead, use the ISO C++ conformant name: _strdup. See online help for details.

if you use `strdup`.

Generally I think it's good to stick to standards when developing cross-platform software. And while `strdup` could land in the next C standard, it's not even in C11 yet.

My suggestion: just like there is `pj_malloc`, why not write `pj_strdup`?

```cpp
char *pj_strdup(const char *str)
{
    size_t len = strlen(str) + 1;
    char *dup = pj_malloc(len);
    if (dup)
        memcpy(dup, str, len);
    return dup;
}
```

**Comment by busstoptaktik** [comment]
I agree, and I actually thought @kbevers implemented it recently, but apparently not.

Not including strdup is certainly one of the more brain-dead decisions of the ISO C89 committee, and I was a bit surprised, reviewing your recent PR, to see how many places it was actually (still) used.

**Comment by kbevers** [comment]
> I actually thought @kbevers implemented it recently, but apparently not.

I did `pj_strlcpy` recently.

`pj_strdup` seems like a good solution to me.

**Comment by busstoptaktik** [comment]
@kbevers: I obviously agree. @rouault - despite your initial, and totally reasonable, comment: would you find it completely unreasonable if @aaronpuchert worked around the MSVC madness by implementing this as part of the PR affected?

**Comment by rouault** [comment]
I'm OK with a pj_strdup

**Comment by busstoptaktik** [comment]
Great, @rouault! And to @aaronpuchert: I believe this is a clear indication that adding pj_strcpy in your PR would be a good way of solving the MSVC CI-problems, while making PROJ.4 slightly more ANSI C89 compliant.

**Comment by aaronpuchert** [comment]
I added commit 6fec8725ab7f9edbbfc8c57c595aa07a297f1f10 to #606. Let's see if this makes MSVC happy.

**Comment by kbevers** [comment]
Fixed with #606
